packages/expo-server/src/vendor/environment/common.ts

```diff
@@ -26,20 +26,76 @@ function initManifestRegExp(manifest: RawManifest): Manifest {
   };
 }
 
+export interface RenderOptions {
+  loader?: { data: unknown };
+}
+
+export interface AssetManifest {
+  js: string[];
+  css: string[];
+}
+
 interface EnvironmentInput {
   readText(request: string): Promise<string | null>;
   readJson(request: string): Promise<unknown>;
   loadModule(request: string): Promise<unknown>;
 }
 
 export function createEnvironment(input: EnvironmentInput) {
+  // Lazy SSR renderer detection, determined on first request
+  // NOTE(@hassankhan): Maybe we could declare this in `routes.json` so we know ahead of time if a
+  // server should be in SSG/SSR mode
+  let ssrRenderer:
+    | ((request: Request, options?: RenderOptions) => Promise<string>)
+    | null
+    | undefined;
+
+  async function getServerRenderer() {
+    if (ssrRenderer === undefined) {
+      try {
+        const mod = await input.loadModule('_expo/server/render.js');
+        const assets = (await input.readJson('_expo/assets.json')) as AssetManifest | null;
```

Comment:
Might be a bit suspect to use `_expo/asset.json` here, as per other comment, but not blocking